Expected type java.lang.Long for property id but imported type is java.lang.Object

Flow 1.0.0.alpha3: I'd like to transfer the following class via Flow:
```java
public interface Entity<ID> {
    ID getId();
}

public class Category implements Serializable, Entity<Long> {
   ...
}
```
(The `Category` class is actually taken from the Beverage Buddy example; I just added the `Entity` interface). When I launch this app, Flow complains with:
```
07-Oct-2017 13:22:11.380 SEVERE [http-nio-8080-exec-3] com.vaadin.server.DefaultErrorHandler.error 
 java.lang.IllegalArgumentException: Unable to create an instance of 'com.vaadin.starter.beveragebuddy.ui.ReviewsList'. The constructor threw an exception.
	at com.vaadin.util.ReflectTools.createProxyInstance(ReflectTools.java:502)
	at com.vaadin.util.ReflectTools.createInstance(ReflectTools.java:439)
	at com.vaadin.flow.di.DefaultInstantiator.createRouteTarget(DefaultInstantiator.java:75)
	at com.vaadin.router.NavigationStateRenderer.lambda$getRouteTarget$99(NavigationStateRenderer.java:87)
	at java.util.Optional.orElseGet(Optional.java:267)
	at com.vaadin.router.NavigationStateRenderer.getRouteTarget(NavigationStateRenderer.java:86)
	at com.vaadin.router.NavigationStateRenderer.handle(NavigationStateRenderer.java:112)
	at com.vaadin.router.Router.navigate(Router.java:107)
	at com.vaadin.router.Router.initializeUI(Router.java:86)
	at com.vaadin.ui.UI.doInit(UI.java:194)
	at com.vaadin.server.BootstrapHandler.createAndInitUI(BootstrapHandler.java:754)
	at com.vaadin.server.BootstrapHandler.synchronizedHandleRequest(BootstrapHandler.java:246)
	at com.vaadin.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1455)
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:298)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:729)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:230)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:192)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:165)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:108)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:472)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:620)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:349)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:784)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:802)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1410)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.vaadin.util.ReflectTools.createProxyInstance(ReflectTools.java:471)
	... 37 more
Caused by: java.lang.IllegalArgumentException: Expected type java.lang.Long for property id but imported type is java.lang.Object
	at com.vaadin.flow.model.BeanModelType.lambda$findBeanGetters$89(BeanModelType.java:520)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
	at com.vaadin.flow.model.BeanModelType.findBeanGetters(BeanModelType.java:511)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
	at com.vaadin.flow.util.ReflectionCache.get(ReflectionCache.java:69)
	at com.vaadin.flow.model.BeanModelType.importProperties(BeanModelType.java:386)
	at com.vaadin.flow.model.BeanModelType.applicationToModel(BeanModelType.java:356)
	at com.vaadin.flow.model.BeanModelType.applicationToModel(BeanModelType.java:46)
	at com.vaadin.flow.model.BeanModelType.lambda$importProperties$86(BeanModelType.java:403)
	at java.util.HashMap.forEach(HashMap.java:1288)
	at com.vaadin.flow.model.BeanModelType.importProperties(BeanModelType.java:401)
	at com.vaadin.flow.model.BeanModelType.applicationToModel(BeanModelType.java:356)
	at com.vaadin.flow.model.ListModelType.importBeans(ListModelType.java:135)
	at com.vaadin.flow.model.ListModelType.applicationToModel(ListModelType.java:90)
	at com.vaadin.flow.model.ListModelType.applicationToModel(ListModelType.java:41)
	at com.vaadin.flow.model.TemplateModelProxyHandler.handleSetter(TemplateModelProxyHandler.java:321)
	at com.vaadin.flow.model.TemplateModelProxyHandler.intercept(TemplateModelProxyHandler.java:166)
	at com.vaadin.starter.beveragebuddy.ui.ReviewsList.ReviewsModel$.setReviews(Unknown Source)
	at com.vaadin.starter.beveragebuddy.ui.ReviewsList.updateList(ReviewsList.java:108)
	at com.vaadin.starter.beveragebuddy.ui.ReviewsList.<init>(ReviewsList.java:78)
	... 42 more
```

However, the field's type is `java.lang.Long`:
```java
System.out.println("RETURN TYPE: " + Category.class.getDeclaredMethod("getId").getReturnType());
```
prints
```
RETURN TYPE: class java.lang.Long
```

--- a/flow-server/src/main/java/com/vaadin/flow/internal/ReflectTools.java
+++ b/flow-server/src/main/java/com/vaadin/flow/internal/ReflectTools.java
@@ -29,13 +29,14 @@ import java.lang.reflect.Type;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
+import java.util.function.Predicate;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import com.googlecode.gentyref.GenericTypeReflector;
 import org.slf4j.LoggerFactory;
 
+import com.googlecode.gentyref.GenericTypeReflector;
 import com.vaadin.flow.shared.util.SharedUtil;
 
 /**
@@ -59,6 +60,8 @@ public class ReflectTools implements Serializable {
     static final String CREATE_INSTANCE_FAILED_NO_PUBLIC_NOARG_CONSTRUCTOR = "Unable to create an instance of '%s'. Make sure the class has a public no-arg constructor.";
     static final String CREATE_INSTANCE_FAILED_LOCAL_CLASS = "Cannot instantiate local class '%s'. Move class declaration outside the method.";
     static final String CREATE_INSTANCE_FAILED_CONSTRUCTOR_THREW_EXCEPTION = "Unable to create an instance of '%s'. The constructor threw an exception.";
+
+    private static final Predicate<Method> IS_SYNTHETIC = Method::isSynthetic;
 
     /**
      * Locates the method in the given class. Returns null if the method is not
@@ -354,7 +357,8 @@ public class ReflectTools implements Serializable {
      * @return a stream of getter methods
      */
     public static Stream<Method> getGetterMethods(Class<?> type) {
-        return Stream.of(type.getMethods()).filter(ReflectTools::isGetter)
+        return Stream.of(type.getMethods()).filter(IS_SYNTHETIC.negate())
+                .filter(ReflectTools::isGetter)
                 .filter(ReflectTools::isNotObjectMethod);
     }
 
@@ -366,7 +370,8 @@ public class ReflectTools implements Serializable {
      * @return a stream of setter methods
      */
     public static Stream<Method> getSetterMethods(Class<?> type) {
-        return Stream.of(type.getMethods()).filter(ReflectTools::isSetter);
+        return Stream.of(type.getMethods()).filter(IS_SYNTHETIC.negate())
+                .filter(ReflectTools::isSetter);
     }
 
     /**

--- a/flow-server/src/test/java/com/vaadin/flow/internal/ReflectToolsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/internal/ReflectToolsTest.java
@@ -15,14 +15,20 @@
  */
 package com.vaadin.flow.internal;
 
-import net.bytebuddy.ByteBuddy;
-import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
+import static org.junit.Assert.assertSame;
+
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.stream.Collectors;
+
 import org.junit.Assert;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
 
-import static org.junit.Assert.assertSame;
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
 
 public class ReflectToolsTest {
 
@@ -70,6 +76,24 @@ public class ReflectToolsTest {
 
         public VarArgsCtor(String... args) {
 
+        }
+    }
+
+    public interface Entity<ID> {
+        ID getId();
+
+        void setId(ID id);
+    }
+
+    public class Category implements Serializable, Entity<Long> {
+
+        @Override
+        public Long getId() {
+            return null;
+        }
+
+        @Override
+        public void setId(Long id) {
         }
     }
 
@@ -226,6 +250,15 @@ public class ReflectToolsTest {
         thrown.expect(IllegalArgumentException.class);
         thrown.expectMessage("a cannot be a primitive type");
         ReflectTools.findCommonBaseType(int.class, Object.class);
+    }
+
+    @Test
+    public void getSetters_classIsGeneric_syntheticMethodsAreFilteredOut() {
+        List<Method> setters = ReflectTools.getSetterMethods(Category.class)
+                .collect(Collectors.toList());
+        Assert.assertEquals(1, setters.size());
+        Method setter = setters.get(0);
+        Assert.assertEquals("setId", setter.getName());
     }
 
     private Class<?> createProxyClass(Class<?> originalClass) {

--- a/flow-server/src/test/java/com/vaadin/flow/internal/ReflectToolsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/internal/ReflectToolsTest.java
@@ -259,6 +259,7 @@ public class ReflectToolsTest {
         Assert.assertEquals(1, setters.size());
         Method setter = setters.get(0);
         Assert.assertEquals("setId", setter.getName());
+        Assert.assertEquals(Long.class, setter.getParameterTypes()[0]);
     }
 
     private Class<?> createProxyClass(Class<?> originalClass) {

